resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala

```diff
@@ -260,16 +267,22 @@ private[spark] class BasicExecutorFeatureStep(
         .withUid(pod.getMetadata.getUid)
         .build()
     }
+
+    val policy = kubernetesConf.get(KUBERNETES_ALLOCATION_PODSALLOCATOR) match {
+      case "statefulset" => "Always"
```

Comment:
Well, I know that StatefulSet requires this, but does this work robustly on the loss of broadcasted variables?

Reply:
Not yet. I'm going to take a look at that after I sort out the PV integration test.

Reply:
So I did some thinking about this (still working on the PV integration test but the code-compile-test cycle is slow), looking at `TorrentBroadcast.scala & `BroadcastManager.scala` if an executor is restarted we'll come back up with an empty `cachedValues` and then if there are any broadcast variables referenced we'll just end up fetching them anew.

We would then run into problems because we'll end up trying to store the block and the disk block manager might say it contains the block when go to do the write.

There are a few ways I can think of addressing this

1) Have the reads "fall through" (e.g. check to see if there present on disk regardless of what we know about in the block manager)
1) Ignore existing file when putting a broadcast block and just hard-write it
1) On startup (in the shell script for the dockerfile) cleanup the contents of the disk blocks location

Personally I'm tempted to do the cleanup approach, since we might have bounced for an OOM or something and I don't want to read back a partially written block. WDYT @dongjoon-hyun ?

Reply:
I was doing more thinking about this and I think #2 (ignore existing and overwrite) is probably the most reasonable, but #1 could also be ok if we changed the behaviour of the DiskStore to do temporary file writes w/atomic renames on write completion.

Personally I'm in favour of going for #2 (ignore existing and overwrite) and leaving the temp file write w/atomic rename as a future improvement.

Reply:
Created https://issues.apache.org/jira/browse/SPARK-36330 to track the potential follow-up.

Reply:
I agree with you with option 3.
> 3. On startup (in the shell script for the dockerfile) cleanup the contents of the disk blocks location
> Personally I'm tempted to do the cleanup approach

Reply:
BTW, `SPARK_LOCAL_DIRS` already exists? Is there a chance which the block managers pick up new storage directory when they are restarted?

Reply:
Ok, I'll update the startup script :)

Reply:
As discussed off-line I switched the solution here to just avoid reusing the execids so we don't need to worry about that.